crates/events/src/lib.rs

```diff
@@ -0,0 +1,594 @@
+    /// let listener = EventListener::new_with_options(&target, "click", options, move |event| {
+    ///     // Stop the event from bubbling
+    ///     event.stop_propagation();
+    ///
+    ///     // ...
+    /// });
+    /// ```
+    #[inline]
+    pub fn new_with_options<S, F>(
+        target: &EventTarget,
+        event_type: S,
+        options: EventListenerOptions,
+        callback: F,
+    ) -> Self
+    where
+        S: Into<Cow<'static, str>>,
+        F: FnMut(Event) + 'static,
+    {
+        let callback = Closure::wrap(Box::new(callback) as Box<FnMut(Event)>);
+
+        Self::raw_new(
+            target,
+            event_type.into(),
+            callback,
+            &options.to_js(false),
+            options.phase,
+        )
+    }
+
+    /// This is exactly the same as [`EventListener::new_with_options`](#method.new_with_options), except the event will only fire once,
+    /// and it accepts `FnOnce` instead of `FnMut`.
+    ///
+    /// It is recommended to use [`EventListener::once`](#method.once) instead, because it has better performance, and it is more convenient.
+    ///
+    /// # Examples
+    ///
+    /// Registers a [`"load"`](https://developer.mozilla.org/en-US/docs/Web/API/FileReader/load_event)
+    /// event and uses
+    /// [`event.prevent_default()`](https://rustwasm.github.io/wasm-bindgen/api/web_sys/struct.Event.html#method.prevent_default):
+    ///
+    /// ```rust,no_run
+    /// # use gloo_events::{EventListener, EventListenerOptions};
+    /// # let target = unimplemented!();
+    /// #
+    /// let options = EventListenerOptions::enable_prevent_default();
+    ///
+    /// let listener = EventListener::once_with_options(&target, "load", options, move |event| {
+    ///     event.prevent_default();
+    ///
+    ///     // ...
+    /// });
+    /// ```
+    ///
+    /// Registers a [`"click"`](https://developer.mozilla.org/en-US/docs/Web/API/Element/click_event)
+    /// event in the capturing phase and uses
+    /// [`event.stop_propagation()`](https://rustwasm.github.io/wasm-bindgen/api/web_sys/struct.Event.html#method.stop_propagation)
+    /// to stop the event from bubbling:
+    ///
+    /// ```rust,no_run
+    /// # use gloo_events::{EventListener, EventListenerOptions};
+    /// # let target = unimplemented!();
+    /// #
+    /// let options = EventListenerOptions::run_in_capture_phase();
+    ///
+    /// let listener = EventListener::once_with_options(&target, "click", options, move |event| {
+    ///     // Stop the event from bubbling
+    ///     event.stop_propagation();
+    ///
+    ///     // ...
+    /// });
+    /// ```
+    #[inline]
+    pub fn once_with_options<S, F>(
+        target: &EventTarget,
+        event_type: S,
+        options: EventListenerOptions,
+        callback: F,
+    ) -> Self
+    where
+        S: Into<Cow<'static, str>>,
+        F: FnOnce(Event) + 'static,
+    {
+        let callback = Closure::once(callback);
+
+        Self::raw_new(
```

Comment:
This doesn't actually ensure that the options have `once: true`, so the listener is not automatically unregistered and if the event is fired more than once it will panic. For example, the `"click"` listener in the above doc comment could fall into this trap.

I think we should either assert that `once: true` is set, force set it ourselves, or otherwise avoid this foot gun.

Reply:
@fitzgen Perhaps you are misunderstanding the code?

`once` is set in the `EventListenerOptions::to_js` method.

The `once_with_options` method calls `to_js(true)`, therefore `once` will always be `true`.

Remember, I *removed* `once` from `EventListenerOptions`, because now the "once-ness" is handled by `once_with_options` vs `new_with_options`

Also note that there is a unit test which does verify that `once_with_options` will only trigger the callback once.